src/main_nep/nep4.cu

```diff
@@ -229,122 +232,88 @@ static void __global__ find_max_min(const int N, const float* g_q, float* g_q_sc
   }
 }
 
-static __global__ void apply_q_scaler(
-  const int N,
-  const NEP4::ANN annmb,
-  const float* __restrict__ g_q_scaler,
-  float* __restrict__ g_descriptors)
+static __global__ void
+apply_q_scaler(const int N, const NEP4::ANN ann, const float* __restrict__ g_q_scaler, float* g_q)
 {
   int n1 = threadIdx.x + blockIdx.x * blockDim.x;
   if (n1 < N) {
-    // Scale descriptors in place
-    for (int d = 0; d < annmb.dim; ++d) {
-      g_descriptors[n1 + d * N] *= g_q_scaler[d];
+    for (int d = 0; d < ann.dim; ++d) {
+      g_q[n1 + d * N] *= g_q_scaler[d];
     }
   }
 }
 
+// Apply the message passing aggregation between neighbors, A*q_theta
 static __global__ void apply_gnn_message_passing(
   const int N,
-  const NEP4::ParaMB paramb,
-  const NEP4::ANN annmb,
-  const NEP4::GNN gnnmb,
+  const NEP4::Para para,
+  const NEP4::ANN ann,
   const float* __restrict__ g_x12,
   const float* __restrict__ g_y12,
   const float* __restrict__ g_z12,
   const float* __restrict__ g_messages,
   const int* g_NN,
   const int* g_NL,
-  float* __restrict__ gnn_descriptors)
+  float* gnn_descriptors)
 {
-  /* Apply the message passing aggregation between neighbors, A*q_theta */
   int n1 = threadIdx.x + blockIdx.x * blockDim.x;
   if (n1 < N) {
-    int neighbor_number = g_NN[n1]; // Number of neighbors to atom i
-    // get messages for atom i and neighbors
-    float q_theta_i[MAX_DIM] = {0.0f};
-    float q_theta_j[MAX_NEIGHBORS * MAX_DIM] = {0.0f}; // maximum size when all atoms are neighbors
-    float fc_ij[MAX_NEIGHBORS] = {0.0f};
-    for (int d = 0; d < annmb.dim; ++d) {
-      q_theta_i[d] = g_messages[n1 + d * N];
-    }
-    for (int j = 0; j < neighbor_number; ++j) {
-      int index = n1 + N * j;
-      int n2 = g_NL[index];
-      for (int d = 0; d < annmb.dim; ++d) {
-        q_theta_j[j + d * neighbor_number] = g_messages[n2 + d * N];
+    int num_neighbors_of_n1 = g_NN[n1];
+    const int F = ann.dim; // dimension of q_out, for now dim_out = dim_in.
+    for (int nu = 0; nu < F; nu++) {
+      float q_out_nu = g_messages[n1 + nu * N]; // fc(r_ii) = 1
+      // TODO perhaps normalize weights? Compare Kipf, Welling et al. (2016)
+      for (int j = 0; j < num_neighbors_of_n1; ++j) {
+        int index = n1 + N * j;
+        int n2 = g_NL[index];
+        float r12[3] = {g_x12[index], g_y12[index], g_z12[index]};
+        float d12 = sqrt(r12[0] * r12[0] + r12[1] * r12[1] + r12[2] * r12[2]);
+        float fc12;
+        find_fc(para.rc_angular, para.rcinv_angular, d12, fc12);
+        q_out_nu += fc12 * g_messages[n2 + nu * N];
```

Comment:
So you move the kernels I wrote from `nep_utilities.cuh` to `nep4.cu`?

Reply:
Yes, I moved two `__device__` functions in `nep_utilities.cuh` to `main_nep/nep4.cu`. 

The `__device__` functions in `nep_utilities.cuh` are supposed to be called by functions in both `main_nep/nepX.cu` and `force/nepX.cu` (`X` = 3 or 4). 

However, to make the CUDA kernels in `main_nep/nep4.cu` more cunderstandable (all the relevant code in the same function), I think it is better to write all the code in the kernels first and condier re-factoring in the future.